src/main/java/io/th0rgal/oraxen/utils/drops/Drop.java

```diff
@@ -1,135 +1,135 @@
-            drop.setItemMeta(clone);
-        }
-
-        if (frame.getItem().getItemMeta() instanceof PotionMeta potionMeta) {
-            PotionMeta clone = (PotionMeta) drop.getItemMeta().clone();
-            clone.setColor(potionMeta.getColor());
-            drop.setItemMeta(clone);
-        }
-
-        frame.getLocation().getWorld().dropItemNaturally(frame.getLocation(), drop);
-    }
-}
+package io.th0rgal.oraxen.utils.drops;
+
+import io.th0rgal.oraxen.items.OraxenItems;
+import io.th0rgal.oraxen.mechanics.provided.misc.itemtype.ItemTypeMechanic;
+import io.th0rgal.oraxen.mechanics.provided.misc.itemtype.ItemTypeMechanicFactory;
+import org.bukkit.Location;
+import org.bukkit.enchantments.Enchantment;
+import org.bukkit.entity.ItemFrame;
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.inventory.meta.LeatherArmorMeta;
+import org.bukkit.inventory.meta.PotionMeta;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ThreadLocalRandom;
+
+public class Drop {
+
+    private List<String> hierarchy;
+    private final List<Loot> loots;
+    final boolean silktouch;
+    final boolean fortune;
+    private final boolean hasMinimalType;
+    String minimalType;
+    private final List<String> bestTools;
+    final String sourceID;
+
+    public Drop(List<String> hierarchy, List<Loot> loots, boolean silktouch, boolean fortune, String sourceID,
+                String minimalType, List<String> bestTools) {
+        this.hierarchy = hierarchy;
+        this.loots = loots;
+        this.silktouch = silktouch;
+        this.fortune = fortune;
+        this.sourceID = sourceID;
+        hasMinimalType = true;
+        this.minimalType = minimalType;
+        this.bestTools = bestTools;
+    }
+
+    public Drop(List<Loot> loots, boolean silktouch, boolean fortune, String sourceID) {
+        this.loots = loots;
+        this.silktouch = silktouch;
+        this.fortune = fortune;
+        this.sourceID = sourceID;
+        hasMinimalType = false;
+        this.bestTools = new ArrayList<>();
+    }
+
+    public String getItemType(ItemStack itemInHand) {
+        String itemID = OraxenItems.getIdByItem(itemInHand);
+        ItemTypeMechanicFactory factory = ItemTypeMechanicFactory.get();
+        if (factory.isNotImplementedIn(itemID)) {
+            String[] content = itemInHand.getType().toString().split("_");
+            return content.length >= 2 ? content[0] : "";
+        } else {
+            ItemTypeMechanic mechanic = (ItemTypeMechanic) factory.getMechanic(itemID);
+            return mechanic.itemType;
+        }
+    }
+
+    public boolean canDrop(ItemStack itemInHand) {
+        return isToolEnough(itemInHand) && isTypeEnough(itemInHand);
+    }
+
+    public boolean isTypeEnough(ItemStack itemInHand) {
+        if (hasMinimalType) {
+            String itemType = itemInHand == null ? "" : getItemType(itemInHand);
+            return !itemType.isEmpty() && hierarchy.contains(itemType)
```

Comment:
iirc isEmpty wouldnt check for a "", for that you would use isBlank which checks for "" or strings with only whitespaces

Reply:
[String.isEmpty()](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/String.html#isEmpty()) Returns true iff length() is 0.

Reply:
guess both works here to isBlank checks for whitespaces aswell which wont be an issue here so